db/dcrpg/pgblockchain.go

```diff
@@ -571,37 +626,91 @@ func (pgb *ChainDB) GetTicketPoolByDateAndInterval(maturityBlock int64,
 	return retrieveTicketsByDate(pgb.db, maturityBlock, int64(val))
 }
 
-// TicketPoolVisualization fetches the following ticketpool data: tickets
+// TicketPoolVisualization helps block consecutive access to the ticketPoolVisualization
+// method till the first request is resolved.
+func (pgb *ChainDB) TicketPoolVisualization(interval dbtypes.ChartGrouping) ([]*dbtypes.PoolTicketsData, *dbtypes.PoolTicketsData, uint64, error) {
+	// Attempt to retrieve data for the current block from cache.
+	heightSeen := pgb.Height()
+	barcharts, donutCharts, height, ok := TicketPoolData(interval, heightSeen)
+	if ok {
+		// The cache was fresh.
+		return barcharts, donutCharts, height, nil
+	}
+
+	// Cache is stale. Attempt to gain updater status.
+	if !pgb.tpUpdatePermission.TryLock() {
+		// Another goroutine is running db query to get the updated data. Return
+		// stale data instead of waiting.
+		return barcharts, donutCharts, height, nil
```

Comment:
I still don't get how this will work because if `ok` is false it means that there is some data missing and its mostly going to barsCharts going by this `barGraphs, isFound = ticketPoolGraphsCache.BarGraphsCache[interval]`....

The so called stale data that you will be serving, most likely you will be serving empty data set except the height.

Reply:
And also this too..
```go
if ticketPoolGraphsCache.Height != height {
		return nil, nil, 0, false
	}
```

Reply:
I've removed the clear function call

Reply:
But if `ok` is false then either `barcharts` or `donutCharts` or both are empty. And after `ok` is false you are not checking if the data you send is empty.

Reply:
This is when the trylock is false.

Reply:
Oh the return nil, nil, ... was a mistake.

This was my error. Should just return the stale data and false. Not nils

Reply:
Just to be on the safe side you also confirm that you will not be sending empty data set when the trylock returns false.